uarray/_uarray_dispatch.cxx

```diff
@@ -0,0 +1,491 @@
+		PyObject * backend, PyObject * args, PyObject * kwargs, PyObject * coerce);
+
+	py_ref canonicalize_args(PyObject * args);
+	py_ref canonicalize_kwargs(PyObject * kwargs);
+
+	static void dealloc(Function * self)
+		{
+			PyObject_GC_UnTrack(self);
+			self->~Function();
+			Py_TYPE(self)->tp_free(self);
+		}
+
+	static PyObject * new_(PyTypeObject * type, PyObject * args, PyObject * kwargs)
+		{
+			auto self = reinterpret_cast<Function *>(type->tp_alloc(type, 0));
+			if (self == nullptr)
+				return nullptr;
+
+			// Placement new
+			self = new (self) Function;
+			return reinterpret_cast<PyObject *>(self);
+		}
+
+	static int init(Function * self, PyObject * args, PyObject * /*kwargs*/)
+		{
+			PyObject * extractor, * replacer;
+			PyObject * backends;
+			PyObject * def_args, * def_kwargs;
+			PyObject * def_impl;
+
+			if (!PyArg_ParseTuple(
+				    args, "OOOO!O!O",
+				    &extractor,
+				    &replacer,
+				    &backends,
+				    &PyTuple_Type, &def_args,
+				    &PyDict_Type, &def_kwargs,
+				    &def_impl))
+			{
+				return -1;
+			}
+
+			if (!PyCallable_Check(extractor) || !PyCallable_Check(replacer))
+			{
+				PyErr_SetString(PyExc_TypeError,
+				                "Argument extractor and replacer must be callable");
+				return -1;
+			}
+
+			if (!PyCallable_Check(backends))
+			{
+				PyErr_SetString(PyExc_TypeError,
+				                "The backends argument must be callable");
+				return -1;
+			}
+
+			if (def_impl != Py_None && !PyCallable_Check(def_impl))
+			{
+				PyErr_SetString(PyExc_TypeError,
+				                "Default implementation must be Callable or None");
+				return -1;
+			}
+
+			self->extractor_ = py_ref::ref(extractor);
+			self->replacer_ = py_ref::ref(replacer);
+			self->backends_ = py_ref::ref(backends);
+			self->def_args_ = py_ref::ref(def_args);
+			self->def_kwargs_ = py_ref::ref(def_kwargs);
+			self->def_impl_ = py_ref::ref(def_impl);
+
+			return 0;
+		}
+};
+
+
+bool is_default(PyObject * value, PyObject * def)
+{
+	// TODO: richer comparison for builtin types? (if cheap)
+	return (value == def);
+}
+
+
+py_ref Function::canonicalize_args(PyObject * args)
+{
+	const auto arg_size = PyTuple_GET_SIZE(args);
+	const auto def_size = PyTuple_GET_SIZE(def_args_.get());
+
+	if (arg_size > def_size)
+		return py_ref::ref(args);
+
+	Py_ssize_t mismatch = 0;
+	for (Py_ssize_t i = arg_size - 1; i >= 0; --i)
+	{
+		auto val = PyTuple_GET_ITEM(args, i);
+		auto def = PyTuple_GET_ITEM(def_args_.get(), i);
+		if (!is_default(val, def))
+		{
+			mismatch = i + 1;
+			break;
+		}
+	}
+
+	return py_ref::steal(PyTuple_GetSlice(args, 0, mismatch));
+}
+
+
+py_ref Function::canonicalize_kwargs(PyObject * kwargs)
+{
+	if (kwargs == nullptr)
+		return py_ref::steal(PyDict_New());
+
+	PyObject * key, * def_value;
+	Py_ssize_t pos = 0;
+	while (PyDict_Next(def_kwargs_, &pos, &key, &def_value))
+	{
+		auto val = PyDict_GetItem(kwargs, key);
+		if (is_default(val, def_value))
```

Comment:
Add `val != nullptr` here?

Reply:
It doesn't fix it and I don't think that should matter since `def_value` shoudn't be `nullptr`.